### ISSUE TITLE
ES modules dynamic import issue

### ISSUE THREAD

**Issue by gacosta89** [thheller/shadow-cljs#1084]
# TLDR;

Seems that dynamic imports are not working, previously it was a thing of the google closure compiler not supporting it but now I think it is supported according to <https://github.com/google/closure-compiler/wiki/JS-Modules#dynamic-import-expressions>.

In this case I am using [Rainbow kit](https://www.rainbowkit.com/) a library to connect and use a crypto wallet. This library is using dynamic imports to fetch svg&rsquo;s <https://github.com/rainbow-me/rainbowkit/blob/2e6bb8ff3850eb4e341d82b77d52b18df4bfd698/packages/rainbowkit/src/wallets/walletConnectors/metaMaskWallet/metaMaskWallet.ts#L53>.

The issue happens when you click the connect button and the main modal appears.

To demonstrate it I created [this demo](https://github.com/revert-finance/shadow-cljs-es-dynamic-imports) that has 2 &ldquo;Connect Wallet&rdquo; buttons side by side.

![img](https://github.com/revert-finance/shadow-cljs-es-dynamic-imports/blob/main/screenshots/home.png?raw=true)

1.  One using the [Rainbow kit](https://www.rainbowkit.com/) npm library as shadow cljs users guide indicates ([using npm packages](https://shadow-cljs.github.io/docs/UsersGuide.html#_using_npm_packages)). [Button source code](https://github.com/revert-finance/shadow-cljs-es-dynamic-imports/blob/main/src/main/rainbow.cljs#L36)
2.  Other using a compiled version of this library that uses a babel plugin to remove the dynamic imports ([babel-plugin-transform-dynamic-imports-to-static-imports](https://github.com/ChristianMurphy/babel-plugin-transform-dynamic-imports-to-static-imports)). Check the vite.config.ts file. [Button source code](https://github.com/revert-finance/shadow-cljs-es-dynamic-imports/blob/main/src/main/rainbow.cljs#L21)

## 1. This is the result when using the library as an npm package

As you can see the wallet icons are missing

![img](https://github.com/revert-finance/shadow-cljs-es-dynamic-imports/blob/main/screenshots/with-dynamic.png?raw=true)

## 2. This is the result when removing the dynamic imports

As you can see the wallet icons are not missing

![img](https://github.com/revert-finance/shadow-cljs-es-dynamic-imports/blob/main/screenshots/without-dynamic.png?raw=true)

**Comment by thheller** [comment]
This is not supported.

The Closure Compiler supported dynamic `import` is still rather buggy and limited. But even if it supported this perfectly, the support for importing `.svg` is still missing. This is not a browser feature, and therefore not supported at all. You can setup webpack as [described here](https://code.thheller.com/blog/shadow-cljs/2020/05/08/how-about-webpack-now.html#option-2-js-provider-external). Or any other bundler the library expects should also work.

**Comment by gacosta89** [comment]
Hi, the npm module is doing a compilation step for svg's making a js file. You can check it in the dist files of the npm module https://www.npmjs.com/package/@rainbow-me/rainbowkit?activeTab=explore in the following path ` dist / wallets / walletConnectors / chunk-DSGPFCE6.js`  line 35

![image](https://user-images.githubusercontent.com/13093156/216391809-83dbdd22-cc9f-46fc-8295-ed42333d0240.png)

And the compiled svg image

![image](https://user-images.githubusercontent.com/13093156/216392538-7e477268-64bf-465f-875d-601f9b85f2f1.png)

So dynamic imports are still not supported in shadow-cljs?

**Comment by jorenbroekema** [comment]
There is also a use case where certain libraries will dynamically import from resources such as a CDN.

For example:

```js
const CDN_URL = `https://ga.jspm.io/npm:${package}@${version}/${entrypoint}`;

const mod = await import(CDN_URL);
```

Bundlers in this case shouldn't even attempt to bundle the dynamic import, it doesn't need to analyze it.

Rollup will just ignore bundling such a dynamic import and give the user a warning, which in this case is desirable.
Webpack and Vite will error unless you wrap the import parameter in an ignore statement:

```js
const CDN_URL = `https://ga.jspm.io/npm:${package}@${version}/${entrypoint}`;

const mod = await import(/* @vite-ignore */ /* webpackIgnore: true */ CDN_URL);
```

I think it would be cool if shadow-cljs could add a way to ignore certain dynamic imports similar to Vite or Webpack, or at least some way to prevent it from throwing an error (e.g. in Rollup you can turn the warning off as well).

Even better of course if this bundler supports dynamic imports but until then this would already be a quick win for dynamic imports that are not intended to be analyzed by bundlers.
